File: app/cost_projection.py
"""Cost projection from USD/1M rates × daily token volume."""

from __future__ import annotations

from typing import Any


PER_1M = 1_000_000


def _round_cost(n: float) -> float:
    return round(float(n) + 0.0, 2)


def project_daily_cost(
    *,
    rate_in_per_1m: float | None,
    rate_out_per_1m: float | None,
    input_tokens_per_day: float | None,
    output_tokens_per_day: float | None,
    team_size: int | float | None = 1,
) -> dict[str, Any] | None:
    """Mirror of ``AppCostProjection.projectDailyCost`` (USD)."""
    rate_in = float(rate_in_per_1m) if rate_in_per_1m is not None else None
    rate_out = float(rate_out_per_1m) if rate_out_per_1m is not None else None
    in_tok = float(input_tokens_per_day or 0)
    out_tok = float(output_tokens_per_day or 0)
    try:
        team = int(team_size) if team_size is not None else 1
    except (TypeError, ValueError):
        team = 1
    team = max(1, team)
    if rate_in is None and rate_out is None:
        return None
    if in_tok <= 0 and out_tok <= 0:
        return None
    day_input = (
        (in_tok / PER_1M) * rate_in * team if rate_in is not None and in_tok > 0 else 0.0
    )
    day_output = (
        (out_tok / PER_1M) * rate_out * team if rate_out is not None and out_tok > 0 else 0.0
    )
    day = day_input + day_output
    if day < 0:
        return None
    return {
        "day_input": _round_cost(day_input),
        "day_output": _round_cost(day_output),
        "day": _round_cost(day),
        "days_7": _round_cost(day * 7),
        "days_30": _round_cost(day * 30),
        "days_365": _round_cost(day * 365),
        "input_tokens_per_day": in_tok,
        "output_tokens_per_day": out_tok,
        "team_size": team,
        "rate_in_per_1m": rate_in,
        "rate_out_per_1m": rate_out,
    }

File: app/static/js/cost-projection.js
/* global window */
/**
 * Project token spend from USD/1M unit rates and daily token volume.
 * Periods: 1 day, 7 days, 30 days, 365 days.
 */
(() => {
  const PER_1M = 1_000_000;

  function _num(v) {
    const n = Number(v);
    return Number.isFinite(n) ? n : null;
  }

  function _roundMoney(n) {
    if (n == null || !Number.isFinite(n)) return null;
    return window.AppMoney?.roundCost?.(n) ?? Math.round(n * 100) / 100;
  }

  /**
   * @param {{
   *   rateInPer1m: number|null|undefined,
   *   rateOutPer1m: number|null|undefined,
   *   inputTokensPerDay: number|null|undefined,
   *   outputTokensPerDay: number|null|undefined,
   * }} opts
   */
  function projectDailyCost({
    rateInPer1m,
    rateOutPer1m,
    inputTokensPerDay,
    outputTokensPerDay,
    teamSize = 1,
  } = {}) {
    const rateIn = _num(rateInPer1m);
    const rateOut = _num(rateOutPer1m);
    const inTok = _num(inputTokensPerDay) ?? 0;
    const outTok = _num(outputTokensPerDay) ?? 0;
    const team = Math.max(1, Math.floor(_num(teamSize) ?? 1));
    if ((rateIn == null && rateOut == null) || (inTok <= 0 && outTok <= 0)) {
      return null;
    }
    const dayInput =
      rateIn != null && inTok > 0 ? (inTok / PER_1M) * rateIn * team : 0;
    const dayOutput =
      rateOut != null && outTok > 0 ? (outTok / PER_1M) * rateOut * team : 0;
    const day = dayInput + dayOutput;
    if (!Number.isFinite(day) || day < 0) return null;
    return {
      day_input: _roundMoney(dayInput),
      day_output: _roundMoney(dayOutput),
      day: _roundMoney(day),
      days_7: _roundMoney(day * 7),
      days_30: _roundMoney(day * 30),
      days_365: _roundMoney(day * 365),
      input_tokens_per_day: inTok,
      output_tokens_per_day: outTok,
      team_size: team,
      rate_in_per_1m: rateIn,
      rate_out_per_1m: rateOut,
    };
  }

  function periodsFromProjection(proj) {
    if (!proj) return [];
    return [
      { key: "day", label: "1 day", days: 1, total: proj.day },
      { key: "week", label: "7 days", days: 7, total: proj.days_7 },
      { key: "month", label: "30 days", days: 30, total: proj.days_30 },
      { key: "year", label: "365 days", days: 365, total: proj.days_365 },
    ];
  }

  window.AppCostProjection = {
    PER_1M,
    projectDailyCost,
    periodsFromProjection,
  };
})();
